sharding/tools/tester.py

```diff
@@ -135,46 +143,89 @@ def __init__(self, alloc=None, env=None):
+        o = self.direct_tx(transaction, shard_id=shard_id)
         self.last_sender = sender
         return o
 
-    def contract(self, sourcecode, args=[], sender=k0, value=0, language='evm', startgas=STARTGAS, gasprice=GASPRICE):
+    def contract(self, sourcecode, args=[], sender=k0, value=0, language='evm', startgas=STARTGAS, gasprice=GASPRICE, shard_id=None):
         if language == 'evm':
             assert len(args) == 0
-            return self.tx(sender=sender, to=b'', value=value, data=sourcecode, startgas=startgas, gasprice=gasprice)
+            return self.tx(sender=sender, to=b'', value=value, data=sourcecode, startgas=startgas, gasprice=gasprice, shard_id=shard_id)
         else:
             compiler = languages[language]
             interface = compiler.mk_full_signature(sourcecode)
             ct = ContractTranslator(interface)
             code = compiler.compile(sourcecode) + (ct.encode_constructor_arguments(args) if args else b'')
-            addr = self.tx(sender=sender, to=b'', value=value, data=code, startgas=startgas, gasprice=gasprice)
-            return ABIContract(self, ct, addr)
+            addr = self.tx(sender=sender, to=b'', value=value, data=code, startgas=startgas, gasprice=gasprice, shard_id=shard_id)
+            return ABIContract(self, ct, addr, shard_id=shard_id)
 
     def mine(self, number_of_blocks=1, coinbase=a0):
         self.cs.finalize(self.head_state, self.block)
         set_execution_results(self.head_state, self.block)
         self.block = Miner(self.block).mine(rounds=100, start_nonce=0)
         assert self.chain.add_block(self.block)
+
+        # Check add_header_logs
+        for item in self.add_header_logs:
+            # [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, bytes]
+            # use sedes to prevent integer 0 from being decoded as b''
+            sedes = List([utils.big_endian_int, utils.big_endian_int, utils.hash32, utils.hash32, utils.hash32, utils.address, utils.hash32, utils.hash32, binary])
+            values = rlp.decode(item, sedes)
+            shard_id = values[0]
+            if shard_id in self.chain.shard_id_list:
+                collation_hash = sha3(item)
+                collation = self.chain.shards[shard_id].get_collation(collation_hash)
+                self.chain.reorganize_head_collation(self.block, collation)
+        # Clear logs
+        self.add_header_logs = []
+
         assert self.head_state.trie.root_hash == self.chain.state.trie.root_hash
         for i in range(1, number_of_blocks):
             b, _ = make_head_candidate(self.chain, timestamp=self.chain.state.timestamp + 14)
             b = Miner(b).mine(rounds=100, start_nonce=0)
             assert self.chain.add_block(b)
         self.block = mk_block_from_prevstate(self.chain, timestamp=self.chain.state.timestamp + 14)
         self.head_state = self.chain.state.ephemeral_clone()
+        self.head_state.log_listeners = self.chain.state.log_listeners
```

Comment:
Because I found that `State.ephemeral_clone()` doesn't clone the `log_listeners` variable:

https://github.com/hwwhww/pyethereum/blob/snapshot_python3/ethereum/state.py#L448-L460

So I just reused it by assigning and I think it's only for tester. (We can use logBloom in daemon.)